### PR TITLE
[scudo] [MTE] resize stack depot for allocation ring buffer

### PR DESCRIPTION
**Pull request by fmayer**

### CODE REVIEW

**Comment by fmayer** [comment]
Need to update error_info_fuzzer, sorry.

File: compiler-rt/lib/scudo/standalone/stack_depot.h [hunk 1]
```diff
@@ -62,47 +63,103 @@ class StackDepot {
...
-  static const uptr RingMask = RingSize - 1;
-  atomic_u64 Ring[RingSize] = {};
+  uptr RingSize = 0;
+  uptr RingMask = 0;
+  uptr TabMask = 0;
+  // This is immediately followed by RingSize atomic_u64 and
+  // (TabMask + 1) atomic_u32.
+
+  atomic_u64 *Ring(char *RawStackDepot) {
+    return reinterpret_cast<atomic_u64 *>(RawStackDepot + sizeof(StackDepot));
```
Comment (cferris1000) [inline compiler-rt/lib/scudo/standalone/stack_depot.h:73]: Would it be better to use the RawDepotMap.getBase() function instead of this? It's already a uptr so no need to do the second cast.

The call is potentially more expensive but I don't think that matters if this is used in debugging contexts.

File: compiler-rt/lib/scudo/standalone/stack_depot.h [hunk 2]
```diff
@@ -62,47 +63,103 @@ class StackDepot {
...
+  }
+
+  atomic_u32 *Tab(char *RawStackDepot) {
+    return reinterpret_cast<atomic_u32 *>(RawStackDepot + sizeof(StackDepot) +
```
Comment (cferris1000) [inline compiler-rt/lib/scudo/standalone/stack_depot.h:77]: Any chance that these values could ever overflow? I know you are planning to allow the setting of the ring buffersize, but since everything is calculated from that value, I don't think an overflow should be possible.

**Review by cferris1000 — COMMENTED** [review]
Is it possible to add some new tests for this? I don't think we have any, so I presume you are using the Android debuggerd tests on a MTE device to verify this.

File: compiler-rt/lib/scudo/standalone/combined.h [hunk 3]
```diff
@@ -1256,7 +1275,8 @@ class Allocator {
   }
 
   void storePrimaryAllocationStackMaybe(const Options &Options, void *Ptr) {
-    if (!UNLIKELY(Options.get(OptionBit::TrackAllocationStacks)))
+    if (!UNLIKELY(Options.get(OptionBit::TrackAllocationStacks)) ||
+        !RawRingBuffer)
```
Comment (ChiaHungDuan) [inline compiler-rt/lib/scudo/standalone/combined.h:1279]: Now the logic of enabling allocation stack becomes slightly complicated. Can we make this logic a function?

File: compiler-rt/lib/scudo/standalone/combined.h [hunk 4]
```diff
@@ -917,8 +919,14 @@ class Allocator {
       Primary.Options.clear(OptionBit::AddLargeAllocationSlack);
   }
 
-  const char *getStackDepotAddress() const {
-    return reinterpret_cast<const char *>(&Depot);
+  const char *getStackDepotAddress() {
+    initThreadMaybe();
```
Comment (ChiaHungDuan) [inline compiler-rt/lib/scudo/standalone/combined.h:923]: This is used to init thread local cache for allocation. Why do we need to call this function here?

File: compiler-rt/lib/scudo/standalone/combined.h [hunk 5]
```diff
@@ -954,45 +962,54 @@ class Allocator {
...
+                           uintptr_t MemoryAddr, size_t MemorySize) {
     *ErrorInfo = {};
     if (!allocatorSupportsMemoryTagging<Config>() ||
         MemoryAddr + MemorySize < MemoryAddr)
       return;
 
-    auto *Depot = reinterpret_cast<const StackDepot *>(DepotPtr);
+    if (DepotPtr && DepotSize < sizeof(StackDepot)) {
+      return;
+    }
```
Comment (ChiaHungDuan) [inline compiler-rt/lib/scudo/standalone/combined.h:988]: lint: Don't need the brackets

And could we have some comment here? Like why we bail out here

File: compiler-rt/lib/scudo/standalone/combined.h [hunk 6]
```diff
@@ -954,45 +962,54 @@ class Allocator {
...
+    if (DepotPtr &&
+        !reinterpret_cast<const StackDepot *>(DepotPtr)->isValid(DepotSize)) {
```
Comment (ChiaHungDuan) [inline compiler-rt/lib/scudo/standalone/combined.h:990]: I would suggest doing
```
// Some comment
if (!DepotPtr)
  return;

// the following logics
```

File: compiler-rt/lib/scudo/standalone/stack_depot.h [hunk 7]
```diff
@@ -62,47 +63,103 @@ class StackDepot {
...
+
+  const atomic_u32 *Tab(const char *RawStackDepot) const {
+    return reinterpret_cast<const atomic_u32 *>(
+        RawStackDepot + sizeof(StackDepot) + sizeof(atomic_u64) * RingSize);
+  }
 
 public:
+  void init(uptr RingSz, uptr TabSz) {
+    DCHECK(isPowerOfTwo(RingSz));
+    DCHECK(isPowerOfTwo(TabSz));
```
Comment (ChiaHungDuan) [inline compiler-rt/lib/scudo/standalone/stack_depot.h:94]: This seems to be checked many times. Maybe we can leave them here and remove the others

File: compiler-rt/lib/scudo/standalone/stack_depot.h [hunk 8]
```diff
@@ -62,47 +63,103 @@ class StackDepot {
...
+    RingMask = RingSz - 1;
+    TabMask = TabSz - 1;
+  }
+
+  // Ensure that RingSize, RingMask and TabMask are set up in a way that
+  // all accesses are within range of BufSize.
+  bool isValid(uptr BufSize) const {
+    if (RingSize > UINTPTR_MAX / sizeof(atomic_u64)) {
+      return false;
+    }
```
Comment (ChiaHungDuan) [inline compiler-rt/lib/scudo/standalone/stack_depot.h:105]: Lint: single statement. here and below

File: compiler-rt/lib/scudo/standalone/stack_depot.h [hunk 9]
```diff
@@ -62,47 +63,103 @@ class StackDepot {
...
+    if (TabMask == 0) {
+      return false;
+    }
+    if ((TabMask - 1) > UINTPTR_MAX / sizeof(atomic_u32)) {
+      return false;
+    }
+    uptr TabSize = TabMask + 1;
+    if (!isPowerOfTwo(TabSize)) {
+      return false;
+    }
```
Comment (ChiaHungDuan) [inline compiler-rt/lib/scudo/standalone/stack_depot.h:123]: If we have asserted that the TabSize should be power of 2, is this still possible to be happened?

**Review by ChiaHungDuan — COMMENTED** [review]
Just took a quick review, will do a second round tomorrow

File: compiler-rt/lib/scudo/standalone/stack_depot.h (hunk 9 above)
Comment (ChiaHungDuan) [reply compiler-rt/lib/scudo/standalone/stack_depot.h:123]: Then do we want to dump any message so that we know why the depot is corrupted? It seems that we only bail out silently.

File: compiler-rt/lib/scudo/standalone/stack_depot.h [hunk 10]
```diff
@@ -62,47 +63,103 @@ class StackDepot {
...
     MurMur2HashBuilder B;
     for (uptr *I = Begin; I != End; ++I)
       B.add(u32(*I) >> 2);
     u32 Hash = B.get();
 
     u32 Pos = Hash & TabMask;
-    u32 RingPos = atomic_load_relaxed(&Tab[Pos]);
-    u64 Entry = atomic_load_relaxed(&Ring[RingPos]);
+    u32 RingPos = atomic_load_relaxed(&Tab(RawStackDepot)[Pos]);
+    u64 Entry = atomic_load_relaxed(&Ring(RawStackDepot)[RingPos]);
```
Comment (ChiaHungDuan) [inline compiler-rt/lib/scudo/standalone/stack_depot.h:151]:
```suggestion
   atomic_u32 *Tab = getTab(RawStackDepot);
   atomic_u64 *Ring = getRing(RawStackDepot);
   u32 RingPos = atomic_load_relaxed(&Tab[Pos]);
   u64 Entry = atomic_load_relaxed(&Ring[RingPos]);
   
   // And the following uses of Tab/Ring
```
`getRing`/`getTab` may be the better names

**Comment by fmayer** [comment]
> Is it possible to add some new tests for this? I don't think we have any, so I presume you are using the Android debuggerd tests on a MTE device to verify this.

Added some tests. Also ran debuggerd tests (seems like I had forgotten to run them on the latest version before, I did run but broke something before uploading - very sorry about that).

File: compiler-rt/lib/scudo/standalone/combined.h [hunk 11]
```diff
@@ -1504,6 +1529,28 @@ class Allocator {
...
+    constexpr auto kFramesPerStack = 8;
+    static_assert(isPowerOfTwo(kFramesPerStack));
+    u32 RingSize = static_cast<u32>(TabSize * kFramesPerStack);
+    DCHECK(isPowerOfTwo(RingSize));
+    static_assert(sizeof(StackDepot) % alignof(atomic_u64) == 0);
+
+    StackDepotSize = sizeof(StackDepot) + sizeof(atomic_u64) * RingSize +
+                     sizeof(atomic_u32) * TabSize;
+    MemMapT DepotMap;
+    DepotMap.map(
```
[the remark this answers is not in the document]
Comment (fmayer) [reply compiler-rt/lib/scudo/standalone/combined.h:1558]: I would prefer to leave this. This will be changed back in a follow up I am working on

File: compiler-rt/lib/scudo/standalone/combined.h [hunk 12]
```diff
@@ -1526,6 +1573,10 @@ class Allocator {
                              RawRingBufferMap.getCapacity());
     }
     RawRingBuffer = nullptr;
+    if (RawStackDepot) {
+      RawStackDepotMap.unmap(RawStackDepotMap.getBase(),
+                             RawStackDepotMap.getCapacity());
+    }
```
[the remark this answers is not in the document]
Comment (fmayer) [reply compiler-rt/lib/scudo/standalone/combined.h:1592]: As above, this is to satisfy the C++ standard rules about object bounds. This is the same we do for the AllocationRingBuffer, and this was discussed and agreed on in that review.

File: compiler-rt/lib/scudo/standalone/stack_depot.h [hunk 13]
```diff
@@ -62,47 +63,103 @@ class StackDepot {
...
+    if (BufSize < RingBytes) {
+      return false;
+    }
+    return BufSize == 0;
+  }
+
   // Insert hash of the stack trace [Begin, End) into the stack depot, and
   // return the hash.
-  u32 insert(uptr *Begin, uptr *End) {
+  u32 insert(char *RawStackDepot, uptr *Begin, uptr *End) {
```
[the remark this answers is not in the document]
Comment (fmayer) [reply compiler-rt/lib/scudo/standalone/stack_depot.h:134]: As above, this is to satisfy the C++ standard rules about object bounds. This is the same we do for the AllocationRingBuffer, and this was discussed and agreed on in that review.

File: compiler-rt/lib/scudo/standalone/combined.h [hunk 14]
```diff
@@ -1504,6 +1529,28 @@ class Allocator {
...
+    StackDepotSize = sizeof(StackDepot) + sizeof(atomic_u64) * RingSize +
+                     sizeof(atomic_u32) * TabSize;
+    MemMapT DepotMap;
+    DepotMap.map(
+        /*Addr=*/0U, roundUp(StackDepotSize, getPageSizeCached()),
+        "scudo:stack_depot");
+    RawStackDepot = reinterpret_cast<char *>(DepotMap.getBase());
+    auto *Depot = reinterpret_cast<StackDepot *>(DepotMap.getBase());
+    Depot->init(RingSize, TabSize);
+    DCHECK(Depot->isValid(StackDepotSize));
```
[the remark this answers is not in the document]
Comment (fmayer) [reply compiler-rt/lib/scudo/standalone/combined.h:1564]: But init doesn't need to know the `StackDepotSize`.

File: compiler-rt/lib/scudo/standalone/stack_depot.h (hunk 9 above)
Comment (fmayer) [reply compiler-rt/lib/scudo/standalone/stack_depot.h]: it seems like a novel concept to have non-fatal errors in scudo. All the logging in report.cpp seems to involve crashing. Happy to add a non-crashing version if desired.

File: compiler-rt/lib/scudo/standalone/combined.h (hunk 3 above)
Comment (fmayer) [reply compiler-rt/lib/scudo/standalone/combined.h:1283]: I don't think that is a good idea. In a follow up I make the RingBuffer atomic, so we need to load it and keep the pointer.

**Comment by fmayer** [comment]
PTAL

File: compiler-rt/lib/scudo/standalone/combined.h [hunk 15]
```diff
@@ -954,45 +963,57 @@ class Allocator {
...
-                           size_t RingBufferSize, const char *Memory,
-                           const char *MemoryTags, uintptr_t MemoryAddr,
-                           size_t MemorySize) {
+                           size_t DepotSize, const char *RegionInfoPtr,
+                           const char *RingBufferPtr, size_t RingBufferSize,
+                           const char *Memory, const char *MemoryTags,
+                           uintptr_t MemoryAddr, size_t MemorySize) {
+    // N.B. we need to support corrupted data in any of the buffers here. We get
+    // this information from an external process (the crashing process) that
+    // shouldn't be able to crash crash_dump.
```
Comment (cferris1000) [inline compiler-rt/lib/scudo/standalone/combined.h:984]: I don't fully understand what this comment means.

What is crash_dump, was that a typo?

File: compiler-rt/lib/scudo/standalone/combined.h [hunk 16]
```diff
@@ -1040,7 +1057,9 @@ class Allocator {
   uptr GuardedAllocSlotSize = 0;
 #endif // GWP_ASAN_HOOKS
 
-  StackDepot Depot;
+  char *RawStackDepot = nullptr;
```
[the remark this answers is not in the document]
Comment (cferris1000) [reply compiler-rt/lib/scudo/standalone/combined.h:1064]: I don't understand what you mean. What out of bounds accesses? You are casting this value nearly everywhere, so I'm confused why not keep it as a pointer to the object.

File: compiler-rt/lib/scudo/standalone/combined.h [hunk 17]
```diff
@@ -1504,6 +1529,28 @@ class Allocator {
...
+    constexpr auto kStacksPerRingBufferEntry = 2;
+    u32 TabSize = static_cast<u32>(roundUpPowerOfTwo(kStacksPerRingBufferEntry *
+                                                     AllocationRingBufferSize));
+    constexpr auto kFramesPerStack = 8;
+    static_assert(isPowerOfTwo(kFramesPerStack));
+    u32 RingSize = static_cast<u32>(TabSize * kFramesPerStack);
+    DCHECK(isPowerOfTwo(RingSize));
+    static_assert(sizeof(StackDepot) % alignof(atomic_u64) == 0);
+
+    StackDepotSize = sizeof(StackDepot) + sizeof(atomic_u64) * RingSize +
```
[the remark this answers is not in the document]
Comment (cferris1000) [reply compiler-rt/lib/scudo/standalone/combined.h:1555]: Ah, I see that now.

File: compiler-rt/lib/scudo/standalone/combined.h (hunk 12 above)
[the remark this answers is not in the document]
Comment (cferris1000) [reply compiler-rt/lib/scudo/standalone/combined.h:1592]: As above, what does this mean? I don't see how this affects object bounds. To be fair, this code is incredibly hard to follow because of the way you are slicing memory and doing the casting and checking.

That's why I would suggest not allocating one big piece, but allocating separate pieces for each individual item. If you are worried about wasting memory, how much memory do you think you would waste? I think maybe it would get you an extra page or two at most. It would make the code easier to follow.

**Review by cferris1000 — CHANGES_REQUESTED** [review]

File: compiler-rt/lib/scudo/standalone/combined.h (hunk 15 above)
Comment (fmayer) [reply compiler-rt/lib/scudo/standalone/combined.h:984]: crash_dump is what consumes this in Android.

File: compiler-rt/lib/scudo/standalone/combined.h (hunk 16 above)
[the remark this answers is not in the document]
Comment (fmayer) [reply compiler-rt/lib/scudo/standalone/combined.h:1064]: We are accessing the entries of the StackDepot through the char* pointer. These entries are past the StackDepot struct, so using a StackDepot* pointer would technically be outside of object bounds. We discussed the same in the CL that made the ring buffer resizable.

File: compiler-rt/lib/scudo/standalone/combined.h (hunk 12 above)
[the remark this answers is not in the document]
Comment (fmayer) [reply compiler-rt/lib/scudo/standalone/combined.h:1592]: > As above, what does this mean? I don't see how this affects object bounds. To be fair, this code is incredibly hard to follow because of the way you are slicing memory and doing the casting and checking.

That's why I would suggest not allocating one big piece, but allocating separate pieces for each individual item. If you are worried about wasting memory, how much memory do you think you would waste? I think maybe it would get you an extra page or two at most. It would make the code easier to follow.

This isn't going to work. This is read by an external process here: https://cs.android.com/android/platform/superproject/main/+/main:system/core/debuggerd/libdebuggerd/scudo.cpp;l=47;drc=cdf55585a788d5b7ad4e4804046549c54d5c1791. This is why this has to be a single contiguous allocation.

File: compiler-rt/lib/scudo/standalone/combined.h (hunk 3 above)
Comment (ChiaHungDuan) [reply compiler-rt/lib/scudo/standalone/combined.h:1283]: I don't understand why the follow up forbids packing the logic to a function?

It can be something like `getRingBuffer` and only returns non-nullptr when it's enabled.

File: compiler-rt/lib/scudo/standalone/combined.h (hunk 14 above)
[the remark this answers is not in the document]
Comment (ChiaHungDuan) [reply compiler-rt/lib/scudo/standalone/combined.h:1564]: It seems to me that it's a logic conflict. Why do we init something with invalid arguments? Shouldn't we stop the initialization before we construct it?

File: compiler-rt/lib/scudo/standalone/combined.h (hunk 16 above)
[the remark this answers is not in the document]
Comment (ChiaHungDuan) [reply compiler-rt/lib/scudo/standalone/combined.h:1064]: I suppose you're referring to the discussion [here](https://reviews.llvm.org/D140932#inline-1361723). 

I wasn't sure what the trailing objects types would be and if it has multiple types, the handling of alignment can be tricky thus I agree with having a quick implementation first. Now I just reviewed the concern and checked the C++ standard, as I mentioned, it's safe to use *this* to locate the address at the end of the object, a pointer is only invalid if the storage life has ended. (And this is how we access the trailing objects)

I think we should just use `StackDeopt` (if you want the struct more self-explained, we can use the trailing array). I'm fine with doing the clean up in a follow up CL if that's easier

File: compiler-rt/lib/scudo/standalone/stack_depot.h (hunk 10 above)
Comment (ChiaHungDuan) [reply compiler-rt/lib/scudo/standalone/stack_depot.h:142]: Is this marked as resolved by accident?

File: compiler-rt/lib/scudo/standalone/combined.h (hunk 14 above)
[the remark this answers is not in the document]
Comment (fmayer) [reply compiler-rt/lib/scudo/standalone/combined.h:1564]: We don't, that's why the `DCHECK` is there? A `DCHECK` should assert what's always true.

[the remark this answers is not in the document]
Comment (ChiaHungDuan) [reply compiler-rt/lib/scudo/standalone/combined.h:1564]: Sorry, I'm not saying how it's verified. I'm talking about the code structure. 

If the `DCHECK` fails, what will you review? It should be the `RingSize` and `TabSize` because they are used for the initialization. Therefore, checking the values before/in the middle of the initialization gives you the useful insight about what's going on. 

Initialization usually involves several steps, it'd be better to tell when there's something wrong instead of carrying the error to the end. If it causes any crash, that DCHECK won't help anything. Besides, at the end of initialization, I would expect that everything is fine. For example, I don't need to verify if `std::vector<int>(4, 0)` gives me 4 elements and all of them are zero.

IMO, the `DCHECK` is not used at the right place.

[the remark this answers is not in the document]
Comment (fmayer) [reply compiler-rt/lib/scudo/standalone/combined.h:1551]: I thought I replied to this, not sure if GitHub problem or I actually didn't:

I don't understand the problem with adding this DCHECK. This makes sure that the logic for `isValid` and for `init` are consistent, which is a good thing to check in debug builds. If it fails it's always something wrong with the code of either of them, because it should not fail by construction of the code.

> I don't need to verify if std::vector<int>(4, 0) gives me 4 elements and all of them are zero.

I don't think that's a correct analogy. If I was the implementer of std::vector then I might want to add some debug checks about the internal state of std::vector.

File: compiler-rt/lib/scudo/standalone/combined.h (hunk 16 above)
[the remark this answers is not in the document]
Comment (fmayer) [reply compiler-rt/lib/scudo/standalone/combined.h]: OK. I removed the raw char* pointers.

File: compiler-rt/lib/scudo/standalone/combined.h (hunk 15 above)
Comment (fmayer) [reply compiler-rt/lib/scudo/standalone/combined.h]: I updated this comment and clarified that crash_dump is talking about the Android use case.

File: compiler-rt/lib/scudo/standalone/combined.h (hunk 3 above)
Comment (fmayer) [reply compiler-rt/lib/scudo/standalone/combined.h:1270]: Sorry, I did this change but messed up a rebase, I'll upload it again shortly

**Comment by fmayer** [comment]
What should I do to drive this forward? I am not sure which comments you want addressed in which way to move forward. I reshuffled the test a bit to make it easier to read by splitting some assertions and the actual implementation. I don't think splitting into more and more mmaps is great as we need to plumb all that through libc and debuggerd in AOSP.

**Review by ChiaHungDuan — COMMENTED** [review]
Sorry for the delay. Most remaining comments are minor style fixes. The last thing I tend to have it fixed is the `DCHECK` for `isValid`. The `isValid should take three arguments, RingSize, TabSize and BufSize and if it's invalid, stop the initialization.

File: compiler-rt/lib/scudo/standalone/combined.h [hunk 18]
```diff
@@ -917,8 +919,14 @@ class Allocator {
...
 
-  const char *getStackDepotAddress() const {
-    return reinterpret_cast<const char *>(&Depot);
+  const char *getStackDepotAddress() {
+    initThreadMaybe();
+    return reinterpret_cast<char *>(Depot);
+  }
+
+  uptr getStackDepotSize() {
+    initThreadMaybe();
```
[the remark this answers is not in the document]
Comment (fmayer) [reply compiler-rt/lib/scudo/standalone/combined.h:928]: `mapAndInitializeRingBuffer` is called in `init` which is called through `initThreadMaybe` (through `initOnceMaybe`).

File: compiler-rt/lib/scudo/standalone/combined.h (hunk 14 above)
[the remark this answers is not in the document]
Comment (fmayer) [reply compiler-rt/lib/scudo/standalone/combined.h:1562]: Discussed offline, there was some misunderstanding, a user should _not_ be able to trigger this. But given we also assert this in tests, I just removed it.

File: compiler-rt/lib/scudo/standalone/combined.h (hunk 18 above)
[the remark this answers is not in the document]
Comment (ChiaHungDuan) [reply compiler-rt/lib/scudo/standalone/combined.h:928]: Got it. Sorry I missed that. Then is calling `mapAndInitializeRingBuffer` enough?

[the remark this answers is not in the document]
Comment (fmayer) [reply compiler-rt/lib/scudo/standalone/combined.h:928]: I am not sure that's a good idea. E.g. if we call `mapAndInitializeRingBuffer` and someone else calls `initThreadMaybe` on another thread, we would have to add synchronization into the `mapAndInitializeRingBuffer`.

[the remark this answers is not in the document]
Comment (ChiaHungDuan) [reply compiler-rt/lib/scudo/standalone/combined.h:929]: Got it. In genereal, `initThreadMaybe` is used when we are going to do some memory allocation/deallocation. Given the `mapAndInitializeRingBuffer` is not synchronized at the function level. I'm fine with doing it like this. I will figure out if we can make it better

**Review by ChiaHungDuan — APPROVED** [review]
